### PR TITLE
Type safe stashing of context values

### PR DESCRIPTION
**Pull request by scothis**
Introducing a Stasher which is a convenience helper for interacting with typed values for a single stash key. The stasher itself is generic and handles casting retrieved values to the desired type.

There are three different ways to handle the value not existing or being of the wrong type:

- RetrieveOrDie panics if a value of the desired type is not found
- RetrieveOrError errors if a value of the desired type is not found
- RetrieveOrEmpty falls back to the empty value for the type

### CODE REVIEW

File: reconcilers/stash.go [hunk 1]
```diff
@@ -55,3 +56,72 @@ func ClearValue(ctx context.Context, key StashKey) interface{} {
...
 }
+
+// Stasher stores and retrieves values from the stash context. The context must be configured
+// with a stash via WithStash(). The stash is pre-configured for the context within a reconciler.
+type Stasher[T any] interface {
+	// Key is the stash key used to store and retrieve the value
+	Key() StashKey
+
+	// Store saves the value in the stash under the key
+	Store(ctx context.Context, value T)
```
Comment (mamachanko) [inline reconcilers/stash.go:67]: **[thought]:** _Store_ is a good name. However, intuitively I was expecting it to be _Stash_. That's in the spirit of _Stringer, Writer, Reader_ etc.

File: reconcilers/stash.go [hunk 2]
```diff
@@ -55,3 +56,72 @@ func ClearValue(ctx context.Context, key StashKey) interface{} {
 	delete(stash, key)
 	return value
 }
+
+// Stasher stores and retrieves values from the stash context. The context must be configured
```
Comment (mamachanko) [inline reconcilers/stash.go:60]: **[suggestion, non-blocking]:** 
```suggestion
// Stasher stores and retrieves values from the stash context. The context which gets passed to its methods must be configured
```

File: reconcilers/stash_test.go [hunk 3]
```diff
@@ -84,10 +84,87 @@ func TestStash_RetrieveValue_UndecoratedContext(t *testing.T) {
...
 	var key StashKey = "stash-key"
 
 	if value := RetrieveValue(ctx, key); value != nil {
 		t.Error("expected RetrieveValue() to return nil for undefined key")
 	}
 }
+
+func TestStasher(t *testing.T) {
+	ctx := WithStash(context.Background())
+	stasher := NewStasher[string]("my-key")
```
Comment (mamachanko) [inline reconcilers/stash_test.go:97]: **[nit]:** I think testing against a stashed value of type `string` is sufficient. Then again maybe the test becomes more interesting when using a non-primitive type like `Example`. In particular I am thinking about `RetrieveOrEmpty`.

**Review by mamachanko — APPROVED** [review]
nice!

File: reconcilers/stash.go (hunk 1 above)
Comment (mamachanko) [reply reconcilers/stash.go:67]: That makes sense to me @scothis! ty

File: reconcilers/stash_test.go (hunk 3 above)
Comment (scothis) [reply reconcilers/stash_test.go:97]: All types in go have a default empty value. There's an example using a vanilla struct here:

https://github.com/reconcilerio/runtime/pull/546/files#diff-21cd712a5bbfcb003ce654296346a4008831cce30760e9b6d2d1fb61a0bb5f4dR235

Comment (mamachanko) [reply reconcilers/stash_test.go:97]: I get it. ty

**Review by mamachanko — APPROVED** [review]

**Pull request merged**